lib/ui/window.dart

```diff
@@ -143,14 +161,234 @@ class Locale {
   /// ```
   ///
   /// The primary language subtag must not be null. The region subtag is
-  /// optional.
-  ///
-  /// The values are _case sensitive_, and should match the case of the relevant
-  /// subtags in the [IANA Language Subtag
-  /// Registry](https://www.iana.org/assignments/language-subtag-registry/language-subtag-registry).
-  /// Typically this means the primary language subtag should be lowercase and
-  /// the region subtag should be uppercase.
-  const Locale(this._languageCode, [ this._countryCode ]) : assert(_languageCode != null);
+  /// optional. The values are _case sensitive_, and must match the values
+  /// listed in
+  /// http://unicode.org/repos/cldr/tags/latest/common/validity/language.xml and
+  /// http://unicode.org/repos/cldr/tags/latest/common/validity/region.xml.
+  ///
+  /// This method only produces standards-compliant instances if valid language
+  /// and country codes are provided. Deprecated subtags will be replaced, but
+  /// incorrectly cased strings are not corrected.
+  const Locale(
+    this._languageCode, [
+    this._countryCode,
+  ]) : assert(_languageCode != null),
+       this._scriptCode = null,
+       this._variants = null,
+       this._extensions = null;
+
+  /// Creates a new Locale object with the specified parts.
+  ///
+  /// This is for internal use only. All fields must already be normalized, must
+  /// already be canonicalized. This method does not modify parameters in any
+  /// way or do any syntax checking.
+  ///
+  /// * language, script and region must be in canonical form.
+  /// * iterating over variants must provide variants in alphabetical order.
+  /// * The extensions map must contain only valid key/value pairs. "u" and "t"
+  ///   keys must be present, with an empty string as value, if there are any
+  ///   subtags for those singletons.
+  Locale._internal(
+    String language, {
+    String script,
```

Comment:
It would be nice if code receiving a locale would be able to reason
about whether that locale is valid, that would help with the problem
of code which inadvertently passes on invalid locales in various
directions.  The two ways which I can imagine this working are:

At runtime:
* an `isValid()` method which can tell you whether this is a locale
  which can be passed to a system understanding BCP 47.  (Maybe
  validity is also a concept that exists at different levels here, but
  for the most restrictive level you'd presumably need a list of
  existing languages and countries).

Through type system:
- `NormalizedLocale` as subclass of Locale where only valid instances
  can be constructed.  (The base class could have a `normalize()`
  method which either works or returns a "can't normalize" error)

Ian, would these approaches work within the Flutter framework's idioms?

Reply:
IMHO we shouldn't care about whether Locales are valid or not, and we shouldn't have logic to check this in the Flutter engine or framework. We can provide that in a package if necessary. In practice, the locales we get will be either from the OS or the developer, and in neither case does failing seem to do the user any good.

Reply:
Taking the "let's _not_ normalize/trust locale instances" idea seriously now, I'll check with other i18n folk experienced with API design what they think.